materials/introduction/02-gem5-basics/02-frequently-used-features/05-generators.py

```diff
@@ -0,0 +1,34 @@
+# generate the first N elements of the Fibonacci sequence
+#   fib[k] = fib[k-1] + fib[k-2]
+def Fibonacci_1(N):
+    fib = [1, 1]
+    for i in range(N - 2):
+        fib.append(fib[-1] + fib[-2])
+    return fib  # return all values at once
+
+
+def Fibonacci_generator(N):
+    a = 0
+    b = 1
+    for count in range(N):
+        a, b = b, a + b
+        yield a
+
+
+if __name__ == "__m5_main__":
+    generator_1 = Fibonacci_generator(10)
+    print(type(generator_1))
+    print("Iterate through all elements")
+    for element in generator_1:
+        print(element)
+    print("Done")
+
+    print("Iterate through all elements again!")
+    for element in generator_1:
+        print(element)
+    print("Done")
+
+    generator_2 = Fibonacci_generator(10)
+    print("Casting a generator to a list")
+    fib_list = list(generator_2)
+    print(fib_list)
```

Comment:
If you want to actually show this example, I suggest doing `fib(100000000)` or something like that. You'll see that the first one takes a very long time (maybe never ends?) but the second does it very quickly.

Reply:
ah, I have an idea, let's do the sum of fib number

Reply:
![image](https://user-images.githubusercontent.com/31742177/176545358-d4beea0b-4ea5-4655-8f1c-3e1573b09a83.png)

Reply:
though, it doesn't feel like I drive home the point here. We are not using the generator only because it's faster.